base/env.jl

```diff
@@ -1,73 +1,64 @@
 # This file is a part of Julia. License is MIT: http://julialang.org/license
 
 @unix_only begin
-    _getenv(var::AbstractString) = ccall(:getenv, Cstring, (Cstring,), var)
-    _hasenv(s::AbstractString) = _getenv(s) != C_NULL
+
+_getenv(var::AbstractString) = ccall(:getenv, Cstring, (Cstring,), var)
+_hasenv(s::AbstractString) = _getenv(s) != C_NULL
+
+function access_env(onError::Function, var::AbstractString)
+    val = _getenv(var)
+    val == C_NULL ? onError(var) : bytestring(val)
+end
+
+function _setenv(var::AbstractString, val::AbstractString, overwrite::Bool=true)
+    ret = ccall(:setenv, Int32, (Cstring,Cstring,Int32), var, val, overwrite)
+    systemerror(:setenv, ret != 0)
+end
+
+function _unsetenv(var::AbstractString)
+    ret = ccall(:unsetenv, Int32, (Cstring,), var)
+    systemerror(:unsetenv, ret != 0)
 end
 
+end # @unix_only
+
 @windows_only begin
+
 const ERROR_ENVVAR_NOT_FOUND = UInt32(203)
-_getenvlen(var::AbstractString) = ccall(:GetEnvironmentVariableW,stdcall,UInt32,(Cwstring,Ptr{UInt8},UInt32),var,C_NULL,0)
-_hasenv(s::AbstractString) = _getenvlen(s)!=0 || Libc.GetLastError()!=ERROR_ENVVAR_NOT_FOUND
-function _jl_win_getenv(s::UTF16String,len::UInt32)
-    val=zeros(UInt16,len)
-    ret=ccall(:GetEnvironmentVariableW,stdcall,UInt32,(Cwstring,Ptr{UInt16},UInt32),s,val,len)
+
+_getenvlen(var::Vector{UInt16}) = ccall(:GetEnvironmentVariableW,stdcall,UInt32,(Ptr{UInt16},Ptr{UInt16},UInt32),var,C_NULL,0)
+_hasenv(s::Vector{UInt16}) = _getenvlen(s) != 0 || Libc.GetLastError() != ERROR_ENVVAR_NOT_FOUND
+_hasenv(s::AbstractString) = _hasenv(cwstring(s))
+
+function access_env(onError::Function, str::AbstractString)
+    var = cwstring(str)
+    len = _getenvlen(var)
+    len == 0 && return Libc.GetLastError() != ERROR_ENVVAR_NOT_FOUND ? utf8("") : onError(str)
```

Comment:
Oh, actually no – I didn't screw it up because of the `return`. But yes, this is obviously confusing.